File: portfolio-pnl-simulator/src/core/workers/simulator.py

import numpy as np
from src.core.models.asset import Asset
from src.core.models.volsurface import VolSurface
from src.core.constant import EXPIRIES, INTEREST_RATE, REGIME_PARAMS, REGIME_TRANSITION, SPX_SPOT, VIX
from src.core.models.market import Market
from src.core.models.regimeparams import RegimeParams
from src.core.models.regime import Regime

def initialize_market() -> Market:
    asset = Asset(symbol='.NDX')
    atm_1m_vol = VIX
    market = Market(
        asset=asset,
        time=-1,
        spot=SPX_SPOT,
        vol_surface=None,
        interest_rate=INTEREST_RATE,
        regime=Regime.CALM
    )
    return simulate_next_market(market, VIX)

def get_next_regime(regime: Regime):    
    probs = REGIME_TRANSITION[regime.value]
    return Regime(np.random.choice(len(probs), p=probs))

def simulate_next_market(market: Market, atm_1m_vol: float = None) -> Market:
    current_time = market.time
    dt = 1/365
    # new expiries
    new_expiries = EXPIRIES - current_time - 1
    new_expiries = new_expiries[new_expiries > 0]
    num_expiries_removed = len(EXPIRIES) - len(new_expiries)
    # remove expired expiries and add new ones at the end with step 7 (weekly)
    expiries_to_add = np.array([new_expiries[-1] + 7 * (i + 1) for i in range(num_expiries_removed)])
    new_expiries = np.concatenate([new_expiries[num_expiries_removed:], expiries_to_add])
    # spot and atm_vol (correlated) shocks
    next_regime = get_next_regime(market.regime)
    regime_params = REGIME_PARAMS[next_regime]
    z1 = np.random.normal()
    z2 = np.random.normal()
    z_spot = z1
    z_vol = regime_params.rho * z1 + np.sqrt(1 - regime_params.rho**2) * z2
    # spot GBM: dS(t) = σS(t)dW --Ito-Calculus--> s(t+1) = s(t) * e^(-σ^2*dt/2 + σ*dW) 
    spot_next = market.spot * np.exp(-0.5 * regime_params.spot_vol**2 * dt + regime_params.spot_vol * np.sqrt(dt) * z_spot)
    # vol Ornstein-Uhlenbeck process
    atm_1m_vol = market.vol_surface.atm_1m_vol_est if atm_1m_vol is None else atm_1m_vol
    atm_1m_vol_next = atm_1m_vol + regime_params.vol_kappa * (regime_params.vol_mean - atm_1m_vol) * dt + regime_params.vol_of_vol * np.sqrt(dt) * z_vol
    vs_next = VolSurface(atm_1m_vol_est=atm_1m_vol_next, expiries=new_expiries, skew=regime_params.skew, convexity=regime_params.convexity, 
                         vol_mean=regime_params.vol_mean, spot=spot_next, interest_rate=market.interest_rate)
    return Market(market.asset, current_time + 1, spot_next, vs_next, market.interest_rate, regime=next_regime)

File: options-market-making/src/core/workers/simulator.cpp
#include "core/workers/simulator.hpp"
#include "core/config.hpp"
#include "core/models/regimeparams.hpp"
#include <random>
#include <cmath>
#include <algorithm>

namespace omm::core::workers {

static std::mt19937 rng(std::random_device{}());
static std::normal_distribution<double> normal(0.0, 1.0);

std::map<omm::core::models::Regime, omm::core::models::RegimeParams> getRegimeParams() {
    std::map<omm::core::models::Regime, omm::core::models::RegimeParams> params;
    
    params[omm::core::models::Regime::CALM] = omm::core::models::RegimeParams(
        0.12,    // spotVol
        0.18,    // volMean
        4.0,     // volKappa
        0.20,    // volOfVol
        -0.3,    // rho
        -0.02,   // skew
        0.01     // convexity
    );
    
    params[omm::core::models::Regime::STRESS] = omm::core::models::RegimeParams(
        0.25,    // spotVol
        0.28,    // volMean
        1.5,     // volKappa
        0.45,    // volOfVol
        -0.6,    // rho
        -0.05,   // skew
        0.02     // convexity
    );
    
    params[omm::core::models::Regime::EVENT] = omm::core::models::RegimeParams(
        0.40,    // spotVol
        0.35,    // volMean
        0.5,     // volKappa
        0.80,    // volOfVol
        -0.75,   // rho
        -0.08,   // skew
        0.03     // convexity
    );
    
    return params;
}

std::shared_ptr<omm::core::models::Market> Simulator::initializeMarket() {
    omm::core::models::Asset asset(".NDX");
    
    auto expiries = Config::getExpiries();
    std::vector<double> expiriesDouble(expiries.begin(), expiries.end());
    
    for (auto& e : expiriesDouble) {
        e += Config::TIME_STEP;
    }
    
    auto volSurface = std::make_shared<omm::core::models::VolSurface>(
        expiriesDouble,
        Config::VIX,
        -0.02,  // skew
        0.01,   // convexity
        0.18,   // volMean
        Config::SPX_SPOT,
        Config::INTEREST_RATE
    );
    
    auto market = std::make_shared<omm::core::models::Market>(
        asset,
        -Config::TIME_STEP,
        Config::SPX_SPOT,
        volSurface,
        Config::INTEREST_RATE,
        omm::core::models::Regime::CALM
    );
    
    auto expiriesForSimulation = Config::getExpiries();
    std::vector<double> expiriesForSimulationDouble(expiriesForSimulation.begin(), expiriesForSimulation.end());
    for (auto& e : expiriesForSimulationDouble) {
        e += Config::TIME_STEP;
    }
    
    return simulateNextMarket(market, Config::VIX, expiriesForSimulationDouble);
}

omm::core::models::Regime Simulator::getNextRegime(omm::core::models::Regime regime) {
    const double* probs = Config::REGIME_TRANSITION[static_cast<int>(regime)];
    
    // Sample from multinomial distribution
    double r = static_cast<double>(rand()) / RAND_MAX;
    double cumProb = 0.0;
    for (int i = 0; i < 3; ++i) {
        cumProb += probs[i];
        if (r < cumProb) {
            return static_cast<omm::core::models::Regime>(i);
        }
    }
    
    return static_cast<omm::core::models::Regime>(2);
}

std::shared_ptr<omm::core::models::Market> Simulator::simulateNextMarket(
    const std::shared_ptr<omm::core::models::Market>& market,
    double atmOneMonthVol,
    const std::vector<double>& expiries
) {
    int currentTime = market->time;
    double dt = static_cast<double>(Config::TIME_STEP) / 365.0;
    
    // New expiries
    std::vector<double> newExpiries;
    if (expiries.empty()) {
        newExpiries = market->volSurface->expiries;
    } else {
        newExpiries = expiries;
    }
    
    // Remove negative expiries
    std::vector<double> filteredExpiries;
    for (double e : newExpiries) {
        if (e - Config::TIME_STEP > 0) {
            filteredExpiries.push_back(e - Config::TIME_STEP);
        }
    }
    
    int numExpiriesRemoved = newExpiries.size() - filteredExpiries.size();
    
    // Add new expiries at the end with 7-day steps
    for (int i = 0; i < numExpiriesRemoved; ++i) {
        double newExpiry = filteredExpiries.back() + 7.0 * (i + 1);
        filteredExpiries.push_back(newExpiry);
    }
    
    // Get next regime and its parameters
    auto nextRegime = getNextRegime(market->regime);
    auto regimeParamsMap = getRegimeParams();
    auto regimeParams = regimeParamsMap[nextRegime];
    
    // Generate correlated shocks
    double z1 = normal(rng);
    double z2 = normal(rng);
    double zSpot = z1;
    double zVol = regimeParams.rho * z1 + std::sqrt(1.0 - regimeParams.rho * regimeParams.rho) * z2;
    
    // Spot GBM
    double spotNext = market->spot * std::exp(
        -0.5 * regimeParams.spotVol * regimeParams.spotVol * dt +
        regimeParams.spotVol * std::sqrt(dt) * zSpot
    );
    
    // Vol Ornstein-Uhlenbeck process
    double currentAtmOneMonthVol = (atmOneMonthVol > 0) ? atmOneMonthVol : market->volSurface->atmOneMonthVolEst;
    double atmOneMonthVolNext = currentAtmOneMonthVol +
        regimeParams.volKappa * (regimeParams.volMean - currentAtmOneMonthVol) * dt +
        regimeParams.volOfVol * std::sqrt(dt) * zVol;
    
    // Create new vol surface
    auto vsNext = std::make_shared<omm::core::models::VolSurface>(
        filteredExpiries,
        atmOneMonthVolNext,
        regimeParams.skew,
        regimeParams.convexity,
        regimeParams.volMean,
        spotNext,
        market->interestRate
    );
    
    return std::make_shared<omm::core::models::Market>(
        market->asset,
        currentTime + Config::TIME_STEP,
        spotNext,
        vsNext,
        market->interestRate,
        nextRegime
    );
}

}  // namespace omm::core::workers
